pkg/prometheus/agent/daemonset.go

```diff
@@ -0,0 +1,319 @@
+	p monitoringv1.PrometheusInterface,
+	c *prompkg.Config,
+	cg *prompkg.ConfigGenerator,
+	tlsSecrets *operator.ShardedSecret,
+) (*appsv1.DaemonSetSpec, error) {
+	cpf := p.GetCommonPrometheusFields()
+
+	pImagePath, err := operator.BuildImagePath(
+		operator.StringPtrValOrDefault(cpf.Image, ""),
+		operator.StringValOrDefault("", c.PrometheusDefaultBaseImage),
+		operator.StringValOrDefault(cpf.Version, operator.DefaultPrometheusVersion),
+		"",
+		"",
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	cpf.EnableFeatures = append(cpf.EnableFeatures, "agent")
+	promArgs := prompkg.BuildCommonPrometheusArgs(cpf, cg)
+	promArgs = appendAgentArgs(promArgs, cg, cpf.WALCompression)
+
+	var ports []v1.ContainerPort
+	if !cpf.ListenLocal {
+		ports = []v1.ContainerPort{
+			{
+				Name:          cpf.PortName,
+				ContainerPort: 9090,
+				Protocol:      v1.ProtocolTCP,
+			},
+		}
+	}
+
+	volumes, promVolumeMounts, err := prompkg.BuildCommonVolumes(p, tlsSecrets)
+	if err != nil {
+		return nil, err
+	}
+
+	configReloaderVolumeMounts := []v1.VolumeMount{
+		{
+			Name:      "config",
+			MountPath: prompkg.ConfDir,
+		},
+		{
+			Name:      "config-out",
+			MountPath: prompkg.ConfOutDir,
+		},
+	}
+
+	var configReloaderWebConfigFile string
+
+	// Mount web config and web TLS credentials as volumes.
+	// We always mount the web config file for versions greater than 2.24.0.
+	// With this we avoid redeploying prometheus when reconfiguring between
+	// HTTP and HTTPS and vice-versa.
+	webConfigGenerator := cg.WithMinimumVersion("2.24.0")
+	if webConfigGenerator.IsCompatible() {
+		var fields monitoringv1.WebConfigFileFields
+		if cpf.Web != nil {
+			fields = cpf.Web.WebConfigFileFields
+		}
+
+		webConfig, err := webconfig.New(prompkg.WebConfigDir, prompkg.WebConfigSecretName(p), fields)
+		if err != nil {
+			return nil, err
+		}
+
+		confArg, configVol, configMount, err := webConfig.GetMountParameters()
+		if err != nil {
+			return nil, err
+		}
+		promArgs = append(promArgs, confArg)
+		volumes = append(volumes, configVol...)
+		promVolumeMounts = append(promVolumeMounts, configMount...)
+
+		// To avoid breaking users deploying an old version of the config-reloader image.
+		// TODO: remove the if condition after v0.72.0.
+		if cpf.Web != nil {
+			configReloaderWebConfigFile = confArg.Value
+			configReloaderVolumeMounts = append(configReloaderVolumeMounts, configMount...)
+		}
+	} else if cpf.Web != nil {
+		webConfigGenerator.Warn("web.config.file")
+	}
+
+	// The /-/ready handler returns OK only after the TSDB initialization has
+	// completed. The WAL replay can take a significant time for large setups
+	// hence we enable the startup probe with a generous failure threshold (15
+	// minutes) to ensure that the readiness probe only comes into effect once
+	// Prometheus is effectively ready.
+	// We don't want to use the /-/healthy handler here because it returns OK as
+	// soon as the web server is started (irrespective of the WAL replay).
```

Comment:
We can either use the same settings as with statefulset mode or consider that since daemonset doesn't support persistent storage, we can have tighter timeouts.